src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Windows.cs

```diff
@@ -33,58 +34,100 @@ public static class OneShotHashProvider
+
+                if (destination.Length < hashSize)
+                {
+                    throw new CryptographicException();
+                }
 
-                int hashSize;
+                HashUpdateAndFinish(cachedAlgorithmHandle, hashSize, source, destination);
 
-                NTSTATUS ntStatus = Interop.BCrypt.BCryptGetProperty(
-                    cachedAlgorithmHandle,
-                    Interop.BCrypt.BCryptPropertyStrings.BCRYPT_HASH_LENGTH,
-                    &hashSize,
-                    sizeof(int),
-                    out _,
-                    0);
+                return hashSize;
+            }
 
-                if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+            private static unsafe bool HashDataUsingPseudoHandle(string hashAlgorithmId, ReadOnlySpan<byte> source, Span<byte> destination, out int hashSize)
+            {
+                hashSize = default;
+
+                Interop.BCrypt.BCryptAlgPseudoHandle algHandle;
+                int digestSizeInBytes;
+
+                if (hashAlgorithmId == HashAlgorithmNames.MD5)
                 {
-                    throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+                    algHandle = Interop.BCrypt.BCryptAlgPseudoHandle.BCRYPT_MD5_ALG_HANDLE;
+                    digestSizeInBytes = 128 / 8;
+                }
+                else if (hashAlgorithmId == HashAlgorithmNames.SHA1)
+                {
+                    algHandle = Interop.BCrypt.BCryptAlgPseudoHandle.BCRYPT_SHA1_ALG_HANDLE;
+                    digestSizeInBytes = 160 / 8;
+                }
+                else if (hashAlgorithmId == HashAlgorithmNames.SHA256)
+                {
+                    algHandle = Interop.BCrypt.BCryptAlgPseudoHandle.BCRYPT_SHA256_ALG_HANDLE;
+                    digestSizeInBytes = 256 / 8;
+                }
+                else if (hashAlgorithmId == HashAlgorithmNames.SHA384)
+                {
+                    algHandle = Interop.BCrypt.BCryptAlgPseudoHandle.BCRYPT_SHA384_ALG_HANDLE;
+                    digestSizeInBytes = 384 / 8;
+                }
+                else if (hashAlgorithmId == HashAlgorithmNames.SHA512)
+                {
+                    algHandle = Interop.BCrypt.BCryptAlgPseudoHandle.BCRYPT_SHA512_ALG_HANDLE;
+                    digestSizeInBytes = 512 / 8;
+                }
+                else
+                {
+                    Debug.Fail("Unknown hash algorithm.");
+                    return false;
                 }
 
-                if (destination.Length < hashSize)
+                if (destination.Length < digestSizeInBytes)
                 {
-                    throw new CryptographicException();
+                    Debug.Fail("Caller should have checked length.");
+                    return false;
                 }
 
-                if (!s_useCompatOneShot)
+                NTSTATUS ntStatus;
+                fixed (byte* pSrc = &MemoryMarshal.GetReference(source))
```

Comment:
A question: why the change to use `MemoryMarshal.GetReference`?

Reply:
Habit, muscle memory, whatever you want to call it. :)

Reply:
The way via [GetPinnableReference](https://github.com/dotnet/runtime/blob/4a58c28b80022311ef6aa982a7c8975af509ef53/src/libraries/System.Private.CoreLib/src/System/ReadOnlySpan.cs#L250-L256) has an extra length check, that can be avoided here, as we know the span won't be empty. So the [codegen](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgBEBLbAcwDsJcGDmFwBuGvSasASgFdeQ/DBYBJBTCgQADgGUNAN2EwxNCQGYG83NgBmMJqQYAxCBAY0A3jQY+mF4ACeGPYAsgAU0jDYACYA8rwANgE6Wti8ADyBwQB8DLgQslBgMACU7tS+DF4Vlb42HAgw0QxhWTAAVAxaGFAMALx5BUWl3rU+1WNjxADsDO3dUOI1YwC+oz5ry+WVxP5B9gByEVFxicmpGW25+YXFJetVD5VsMMCy3CwAgri4GhhhN2GDAAhANeLIEgkSktJttJvVGs1WvtOgt+gwAGQhGD4aABELYKC4AAW2ASLAA4jAMJE7LBeMUAUM7vctpUJrDfDM5gsYZNNpVNisgA) is better (maybe the reason for Levi's muscle memory to kick in 😉).